examples/vision/fully_convolutional_network.py

```diff
@@ -0,0 +1,630 @@
+INPUT_HEIGHT = 224
+INPUT_WIDTH = 224
+LEARNING_RATE = 1e-3
+WEIGHT_DECAY = 1e-4
+EPOCHS = 10
+BATCH_SIZE = 32
+MIXED_PRECISION = True
+SHUFFLE = True
+
+# Mixed-precision setting
+if MIXED_PRECISION:
+    policy = keras.mixed_precision.Policy("mixed_float16")
+    keras.mixed_precision.set_global_policy(policy)
+
+"""
+## Load dataset
+
+We make use of the [Oxford-IIIT Pets dataset](http://www.robots.ox.ac.uk/~vgg/data/pets/)
+which contains a total of 7,349 samples and their segmentation masks.
+We have 37 classes, with roughly 200 samples per class.
+Our training and validation dataset has 3,128 and 552 samples respectively.
+Aside from this, our test split has a total of 3,669 samples.
+
+We set a `batch_size` parameter that will batch our samples together, use a `shuffle`
+parameter to mix our samples together.
+"""
+
+(train_ds, valid_ds, test_ds) = tfds.load(
+    "oxford_iiit_pet",
+    split=["train[:85%]", "train[85%:]", "test"],
+    batch_size=BATCH_SIZE,
+    shuffle_files=SHUFFLE,
+)
+
+"""
+## Unpack and Pre-process dataset
+
+We define a simple function that includes performs Resizing over our
+training, validation and test datasets.
+We do the same process on the masks as well, to make sure both are aligned in terms of shape and size.
+"""
+
+
+# Image and Mask Pre-processing
+def unpack_resize_data(section):
+    image = section["image"]
+    segmentation_mask = section["segmentation_mask"]
+
+    resize_layer = keras.layers.Resizing(INPUT_HEIGHT, INPUT_WIDTH)
+
+    image = resize_layer(image)
+    segmentation_mask = resize_layer(segmentation_mask)
+
+    return image, segmentation_mask
+
+
+train_ds = train_ds.map(unpack_resize_data, num_parallel_calls=AUTOTUNE)
+valid_ds = valid_ds.map(unpack_resize_data, num_parallel_calls=AUTOTUNE)
+test_ds = test_ds.map(unpack_resize_data, num_parallel_calls=AUTOTUNE)
+"""
+## Visualize one random sample from the pre-processed dataset
+
+We visualize what a random sample in our test split of the dataset looks like, and plot
+the segmentation mask on top to see the effective mask areas.
+Note that we have performed pre-processing on this dataset too,
+which makes the image and mask size same.
+"""
+
+# Select random image and mask. Cast to NumPy array
+# for Matplotlib visualization.
+
+images, masks = next(iter(test_ds))
+random_idx = tf.random.uniform([], minval=0, maxval=BATCH_SIZE, dtype=tf.int32)
+
+test_image = images[random_idx].numpy().astype("float")
+test_mask = masks[random_idx].numpy().astype("float")
+
+# Overlay segmentation mask on top of image.
+fig, ax = plt.subplots(nrows=1, ncols=2, figsize=(10, 5))
+
+ax[0].set_title("Image")
+ax[0].imshow(test_image / 255.0)
+
+ax[1].set_title("Image with segmentation mask overlay")
+ax[1].imshow(test_image / 255.0)
+ax[1].imshow(
+    test_mask,
+    cmap="inferno",
+    alpha=0.6,
+)
+plt.show()
+
+"""
+## Perform VGG-specific pre-processing
+
+`keras.applications.VGG19` requires the use of a `preprocess_input` function that will
```

Comment:
VGG19 is badly outdated. Can you use a more modern backbone?

Reply:
I can, but the FCN model itself mainly uses VGG as the backbone which gave the highest performance. Another thing to note is the fact that most other candidates like ConvNext have specialized blocks present which may defeat the purpose of it being "Fully-convolutional". EfficientNet may still work too, but it still hasn't been something that is tested from the paper. 

We can deviate from the paper though and try EfficientNets since they seem to still be closer to the fully-convolutional architecture